feat(grpc-proxy): add gRPC Server Reflection support (#821)

Tsukikage7 opened the pull request:
feat(grpc-proxy): add gRPC Server Reflection support (#821)

Implement dynamic message parsing capability for gRPC proxy filter,
replacing the hardcoded ForceCodec passthrough mechanism.

New features:
- Three reflection modes: passthrough, reflection, hybrid
- Dynamic message decoding via protobuf reflection (DynamicCodec)
- Protocol detection for gRPC/Triple compatibility
- TTL-based descriptor caching with automatic cleanup
- Triple metadata extraction (tri-* headers)

New files:
- protocol.go: Protocol detection and Triple metadata handling
- descriptor_cache.go: TTL-based method descriptor caching
- reflection_manager.go: gRPC reflection with official protobuf libs
- dynamic_codec.go: Runtime message inspection/manipulation

Test coverage: 31.2% for proxy package

Closes #821

Alanxtl commented:
can u provide a brief demo to `dubbo-go-pixiu-samples` with the purpose of testing the new feature?

Tsukikage7 commented:
> can u provide a brief demo to `dubbo-go-pixiu-samples` with the purpose of testing the new feature?

I agree with providing a brief demo in dubbo-go-pixiu-samples to test the new feature.

I will add a simple sample there and update this PR accordingly.
Thanks for the suggestion.

Similarityoung commented on pkg/filter/network/grpcproxy/filter/proxy/grpc_proxy_filter.go:
```diff
@@ -174,18 +220,74 @@ func (f *Filter) handleStream(ctx *grpcCtx.GrpcContext, address string) filter.F
+	if ct, ok := ctx.Attachments["content-type"].(string); ok {
+		contentType = ct
+	}
+	protocol := DetectProtocol(contentType)
```
这里为什么需要检测类型，这里的内容我记得只会从 grpclistener 里过来，triple 的话应该用的是triplelistener而不是grpclistener；

换句话说，我觉得这里如果是triple的话就不会走这里了。

Similarityoung commented on pkg/filter/network/grpcproxy/filter/proxy/reflection_manager.go:
```diff
@@ -0,0 +1,413 @@
+	// Request file descriptor for the service
+	fileDescs, err := rm.resolveServiceFileDescriptors(stream, serviceName)
+	if err != nil {
+		return nil, err
```
### AI review

`getOrCreateFileRegistry` 方法使用了 rm.mu 包裹了整个 `getOrCreateFileRegistry` 逻辑，包括网络调用 `client.ServerReflectionInfo(ctx)`。

```go
rm.mu.Lock()
defer rm.mu.Unlock()
// ...
stream, err := client.ServerReflectionInfo(ctx) // 网络 IO
```

这是一个全局锁。虽然在 Cache Hit 时不会触发，但在高并发冷启动（大量新服务同时请求）或缓存批量过期时，会导致所有 Reflection 请求（甚至是对不同服务的请求）串行化，严重影响性能。 

建议: 使用 singleflight 模式针对每个 address 进行通过，或者将锁的粒度细化到 address 级别，避免全局阻塞。

Alanxtl commented on pkg/filter/network/grpcproxy/filter/proxy/descriptor_cache.go:
```diff
@@ -0,0 +1,367 @@
+func NewDescriptorCacheWithSize(ttl time.Duration, maxSize int) *DescriptorCache {
+	if maxSize < minCacheSize {
+		maxSize = minCacheSize
+		logger.Infof("Cache size adjusted to minimum: %d", maxSize)
```
add a logger prefix

Alanxtl commented on pkg/filter/network/grpcproxy/filter/proxy/descriptor_cache.go:
```diff
@@ -0,0 +1,367 @@
+type DescriptorCache struct {
+	entries     sync.Map   // key -> *cacheEntry
+	lruList     *list.List // LRU eviction list
+	lruMutex    sync.Mutex // Protects lruList
```
I suggest use gost's lru https://github.com/dubbogo/gost/blob/master/container/gxlru/lru_cache.go

Alanxtl reviewed (commented):
can u help writing a doc about using grpcproxy to `docs/user/filter`?
and sync docs to https://github.com/apache/dubbo-website

Alanxtl commented:
our project use `import-formatter` to format import blocks, that's the reason why ur CI fails. For you, u should 
1. run `go install github.com/dubbogo/tools/cmd/imports-formatter@latest`
2. cd to the root dir of `dubbo-go-pixiu`
3. run `imports-formatter`

Alanxtl commented:
the grpc/reflection intergrate test failed
集成测试里面的config默认只使用conf.yaml不使用其他的config，会是这个原因吗

Tsukikage7 committed:
feat(grpc-proxy): add gRPC Server Reflection support (#821)

Add gRPC reflection mode to enable content-aware proxying:
- Support three modes: passthrough, reflection, hybrid
- Add descriptor caching with TTL and LRU eviction
- Add dynamic codec for message inspection
- Fix unary call deadlock by detecting stream type

Tsukikage7 commented:
> the grpc/reflection intergrate test failed 集成测试里面的config默认只使用conf.yaml不使用其他的config，会是这个原因吗

可能是Unary的死锁bug导致的，我修了一下

Alanxtl commented:
pls don't forget import-format before commit

Alanxtl reviewed (approved):
LGTM
